### PR TITLE
User registration

### LINKED ISSUE #3: User registration
- [x] Create Login Component
- [x] Create Registration Component
- [x] Add NavLink on Register button
- [x] Create App-reducer
- [x] Create Auth-reducer
- [x] Create Container Components
- [x] Create Redux-store 
- [x] Create Utils directory with validators on client side
- [x] Create Utils directory with validators on server side
- [x] Create routes on server side
- [x] Create controllers on server side
- [x] Add jwt token authorization
- [x] Add refresh token
- [x] Add refresh token in cookies
- [x] Set and get tokens in localStorage
- [x] Create roles methods for users
- [x] Create roles associations in db
- [x]  Create redux-saga structure
- [x]  Create saga files for 
- [x]  Create mobile notification functionality on server side 
- [x]  Create axios layer and http requests on client side
- [x]  Add authentification interceptor
- [x]  Make route authentification tracking
- [x]  Control user roles on server and client side
- [x]  Handle Errors on client side
- [x]  Create roleMiddleware for tracking routes
- [x] Create main Navbar for app navigation
- [x] Save state after reload and get user info from token
- [x]  Add refresh token logic

### PR DESCRIPTION
**Pull request by N-Balinskiy**

### CODE REVIEW

File: client/src/App.scss [hunk 1]
```diff
@@ -0,0 +1,150 @@
+body {
+  background-color: #e9ecef !important;
```
Comment (andrii-kuzmichov) [inline client/src/App.scss]: Create constants for colors

File: client/src/App.scss [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  max-width: 600px;
+  margin: auto;
+}
+
+.text-center {
+  text-align: center;
+}
+
+.box {
+  background-color: #ffff;
```
Comment (andrii-kuzmichov) [inline client/src/App.scss]: typo?

File: client/src/App.scss [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  margin: 20px 0 20px 0;
```
Comment (andrii-kuzmichov) [inline client/src/App.scss]: You can use this format `margin: 20px 0;`

File: client/src/App.scss [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+#driver {
+  margin-left: 10px;
+}
+
+.col-xs-4 > span {
+  margin-left: 5px;
+  color: red;
+}
```
Comment (andrii-kuzmichov) [inline client/src/App.scss]: Separate this file to several by functionality

File: client/src/App.tsx [hunk 5]
```diff
@@ -1,21 +1,22 @@
-import store from './reduxStore';
-import React from 'react';
-import 'bootstrap/dist/css/bootstrap.min.css';
 import { BrowserRouter } from 'react-router-dom';
 import { Provider } from 'react-redux';
+import { ConnectedRouter } from 'connected-react-router';
+
```
Comment (andrii-kuzmichov) [inline client/src/App.tsx:4]: Remove empty row

File: client/src/App.tsx [hunk 6]
```diff
@@ -1,21 +1,22 @@
...
 
-/**
- * @return {Object}
- */
+import './App.scss';
+import 'bootstrap/dist/css/bootstrap.min.css';
+import './App.scss';
+import Feedback from './pages/feedback/Feedback';
+
 function App() {
```
Comment (andrii-kuzmichov) [inline client/src/App.tsx]: Replace to arrow func and don't used return

File: client/src/actions/authActions.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { AuthActionTypes, IAuthAction } from '../types/userTypes';
+
+export const registrate = (payload: any): IAuthAction => ({
```
Comment (andrii-kuzmichov) [inline client/src/actions/authActions.ts]: Create interfaces for each action payload

File: client/src/actions/cityInfoActions.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import { CityInfoValues, InfoActionTypes } from '../types/cityInfoTypes';
+import { CityInfoI } from '../types/cityInfoTypes';
+
+export const getCityInfoCreator = (cityName: string) => {
+  return {
```
Comment (andrii-kuzmichov) [inline client/src/actions/cityInfoActions.ts]: Don't use `return` in place there it is possible

File: client/src/components/Alert/Alert.scss [hunk 9]
```diff
@@ -0,0 +1,4 @@
+.toast, .toast-header {
+    background-color: rgb(255, 116, 116) !important;
```
Comment (andrii-kuzmichov) [inline client/src/components/Alert/Alert.scss]: Let use just one type of color, like as Hex format, and try to don't use `important`

File: client/src/components/Alert/Alert.tsx [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import Toast from 'react-bootstrap/Toast';
+import './Alert.scss';
+
+const Alert = (props: any) => {
```
Comment (andrii-kuzmichov) [inline client/src/components/Alert/Alert.tsx:5]: add interface

File: client/src/components/Alert/Alert.tsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  const changeError = () => {
+  props.handleError({data: '', hasError: false});
```
Comment (andrii-kuzmichov) [inline client/src/components/Alert/Alert.tsx]: put the tab(space)

File: client/src/components/Alert/Alert.tsx [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  };
+
+  if (props?.auth?.hasError) {
```
Comment (andrii-kuzmichov) [inline client/src/components/Alert/Alert.tsx]: move this check to return block and replace to ternary operator

File: client/src/components/Alert/AlertContainer.tsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+import { errorHandle } from '../../actions/authActions';
+
+const mapStateToProps = (state: any) => {
+  return {
+    auth: state.auth,
+  };
+};
+
+const mapDispathToProps = (dispatch: any) => (
+    {handleError: (payload: any) => {
```
Comment (andrii-kuzmichov) [inline client/src/components/Alert/AlertContainer.tsx]: Let to do formatter this code

File: client/src/components/CarTypeSelect/CarTypesSelect.tsx [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    <Form.Control
+      as="select"
+      id={id}
+      className="form-select form-control col-xs-4"
+      aria-label="Car type select"
+      value={selectedValue}
+      onChange={onChange}
+    >
+      {carTypes.map(({ id, name }) => {
+        return (
```
Comment (andrii-kuzmichov) [inline client/src/components/CarTypeSelect/CarTypesSelect.tsx:21]: reduce return using

File: client/src/components/ExtraServiceItem.tsx [hunk 15]
```diff
@@ -0,0 +1,71 @@
+import React, { FC } from 'react';
+import { OverlayTrigger, Tooltip, Form } from 'react-bootstrap';
+
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServiceItem.tsx:3]: Remove empty row

File: client/src/components/ExtraServiceItem.tsx [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+import { ReactComponent as BabyChair } from './icons/babyChair.svg';
+import { ReactComponent as En } from './icons/en.svg';
+import { ReactComponent as Silent } from './icons/silent.svg';
+import { FormLabel } from './FormLabel/FormLabel';
+
+interface IconsI {
+  [index: string]: any;
+}
+
+const extraServicesIcons: IconsI = {
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServiceItem.tsx:13]: Move to diff file

File: client/src/components/ExtraServiceItem.tsx [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  onExtraServicesChanged,
+  isActive,
+  showTitle,
+  iconActiveClass = 'active',
+  iconClasses = [],
+}) => {
+  const iconClass = iconClasses;
+  if (isActive) {
+    iconClass.push(iconActiveClass);
+  }
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServiceItem.tsx:40]: Move this logic to container

File: client/src/components/ExtraServices/ExtraServiceItem.tsx [hunk 18]
```diff
@@ -0,0 +1,71 @@
+import React, { FC } from 'react';
+import { OverlayTrigger, Tooltip, Form } from 'react-bootstrap';
+
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServices/ExtraServiceItem.tsx]: Empty row

File: client/src/components/ExtraServices/ExtraServiceItem.tsx [hunk 19]
```diff
@@ -0,0 +1,71 @@
+import React, { FC } from 'react';
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServices/ExtraServiceItem.tsx:1]: Duplicate components

File: client/src/components/ExtraServices/ExtraServicesSelect.tsx [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+  onExtraServicesChanged: () => any;
+}
+
+export const ExtraServicesSelect: FC<ExtraServicesSelectI> = ({
+  title,
+  avaliableInCityExtraServices,
+  activeExtraServices,
+  onExtraServicesChanged,
+}) => {
+  return avaliableInCityExtraServices ? (
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServices/ExtraServicesSelect.tsx:12]: Return

File: client/src/components/ExtraServices/ExtraServicesSelect.tsx [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+        variant="link"
+        eventKey="0"
+      >
+        {title || 'Extra services:'}
+      </Accordion.Toggle>
+      <Card>
+        <Accordion.Collapse eventKey="0">
+          <Card.Body className="extra-services">
+            {avaliableInCityExtraServices.map(({ id, name }) => {
+              const isActive = activeExtraServices.includes(id);
```
Comment (andrii-kuzmichov) [inline client/src/components/ExtraServices/ExtraServicesSelect.tsx:26]: let use to params and don't use return

File: client/src/components/FormLabel/FormLabel.tsx [hunk 22]
```diff
@@ -0,0 +1,26 @@
+import React, { FC } from 'react';
+import { Form } from 'react-bootstrap';
+
```
Comment (andrii-kuzmichov) [inline client/src/components/FormLabel/FormLabel.tsx]: empty row

File: client/src/components/FormLabel/FormLabel.tsx [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+  classNames?: string[];
+}
+
+export const FormLabel: FC<FormLabelI> = ({
+  classNames = [],
+  title,
+  htmlFor,
+  children,
+}) => {
+  const classes = classNames.concat('col-xs-2', 'form__label').join(' ');
```
Comment (andrii-kuzmichov) [inline client/src/components/FormLabel/FormLabel.tsx:12]: why you use concat? Could you pass classNames as string to attr

File: client/src/components/Input/Input.tsx [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+export const Input: FC<InputI> = ({
+  onChange,
+  value,
+  className,
+  type,
+  name,
+  id,
+  placeholder,
+}) => {
+  return (
```
Comment (andrii-kuzmichov) [inline client/src/components/Input/Input.tsx]: reduce return

**Review by andrii-kuzmichov — COMMENTED** [review]
Replace all string and number to constants

File: client/src/components/Input/Input.tsx (hunk 24 above)
Comment (N-Balinskiy) [reply client/src/components/Input/Input.tsx]: End with code correction but I`m not changed anything in map and order components

**Review by andrii-kuzmichov — APPROVED** [review]

**Review by babchukbohdan — APPROVED** [review]

**Review by romanh24 — APPROVED** [review]

**Pull request merged**